### ISSUE TITLE
Current Location Sensitive Loss form based on `torch.where` is not differentiable.

### ISSUE THREAD

**Issue by lartpang** [Lliu666/MSHNet#5]
@Lliu666  Congrats on being accepted by CVPR.

https://github.com/Lliu666/MSHNet/blob/3af39d64c1fb87400eb586b33053fce8bd388f1b/model/loss.py#L77-L102

Is this the actual code used in the paper?
The current loss form based on `torch.where` is not differentiable.
So this doesn't seem to work as emphasized in the paper?

```python
import torch
import torch.nn as nn


def ShapeLoss(pred, target):
    loss = torch.tensor(0.0, requires_grad=True, device=pred.device)
    patch_size = pred.shape[0]
    smooth = 0.001
    for i in range(patch_size):
        _, pred_x, pred_y = torch.where(pred[i] > 0)
        _, target_x, target_y = torch.where(target[i] > 0)

        pred_centerx = torch.sum(pred_x) / pred_x.shape[0]
        pred_centery = torch.sum(pred_y) / pred_y.shape[0]

        target_centerx = torch.sum(target_x) / target_x.shape[0]
        target_centery = torch.sum(target_y) / target_y.shape[0]

        angle_loss = (4 / (torch.pi**2)) * (
            torch.square(
                torch.arctan((pred_centery + smooth) / (pred_centerx + smooth))
                - torch.arctan((target_centery + smooth) / (target_centerx + smooth))
            )
        )

        pred_length = torch.sqrt(torch.pow(pred_centerx, 2) + torch.pow(pred_centery, 2))
        target_length = torch.sqrt(torch.pow(target_centerx, 2) + torch.pow(target_centery, 2))

        length_loss = (torch.min(pred_length, target_length) + smooth) / (
            torch.max(pred_length, target_length) + smooth
        )
        loss = loss + (1 - length_loss.long() + angle_loss.long()) / patch_size

    return loss


class SLSIoULoss(nn.Module):
    def __init__(self):
        super(SLSIoULoss, self).__init__()

    def forward(self, pred, target, warm_epoch=0, epoch=1, layer=0, with_shape=False):
        pred = torch.sigmoid(pred)
        smooth = 0.0001

        intersection = pred * target

        intersection_sum = torch.sum(intersection, dim=(1, 2, 3))
        pred_sum = torch.sum(pred, dim=(1, 2, 3))
        target_sum = torch.sum(target, dim=(1, 2, 3))

        dis = torch.pow((pred_sum - target_sum) / 2, 2)

        if epoch > warm_epoch:
            alpha = (torch.min(pred_sum, target_sum) + dis + smooth) / (
                torch.max(pred_sum, target_sum) + dis + smooth
            )
            loss = (alpha) * (intersection_sum + smooth) / (pred_sum + target_sum - intersection_sum + smooth)

            if with_shape:
                loss = 1 - loss.mean() + ShapeLoss(pred, target)
            else:
                loss = 1 - loss.mean()
        else:
            loss = (intersection_sum + smooth) / (pred_sum + target_sum - intersection_sum + smooth)

            loss = 1 - loss.mean()
        return loss


if __name__ == "__main__":
    torch.manual_seed(2048)

    pred = torch.rand(1, 1, 5, 5, requires_grad=True)
    target = torch.randint(low=0, high=2, size=(1, 1, 5, 5))
    sls_iou_loss = SLSIoULoss()

    loss = sls_iou_loss(pred, target, with_shape=True)  # original form
    loss.backward()
    print(pred.grad)

    pred.grad.zero_()
    loss = sls_iou_loss(pred, target, with_shape=False)
    loss.backward()
    print(pred.grad)
```

output:

```text
$ python .\test_pytorch.py
1.13.1+cpu
tensor([[[[-0.0071,  0.0035, -0.0088,  0.0037,  0.0040],
          [-0.0080, -0.0078,  0.0042, -0.0077, -0.0088],
          [-0.0087,  0.0036,  0.0039, -0.0088,  0.0037],
          [ 0.0041,  0.0039, -0.0073,  0.0038,  0.0043],
          [-0.0081,  0.0040,  0.0034, -0.0073,  0.0036]]]])
tensor([[[[-0.0071,  0.0035, -0.0088,  0.0037,  0.0040],
          [-0.0080, -0.0078,  0.0042, -0.0077, -0.0088],
          [-0.0087,  0.0036,  0.0039, -0.0088,  0.0037],
          [ 0.0041,  0.0039, -0.0073,  0.0038,  0.0043],
          [-0.0081,  0.0040,  0.0034, -0.0073,  0.0036]]]])
```

**Comment by Lliu666** [comment]
Thanks for your interest in our work and for pointing out the issue for us. This repo. is a reimplementation. The actual code used in the paper is written by another guy, which is ugly and not easy to read. After talking with him, I have fixed the bug, please have a try.
